From 8f1734383c2cb5f88d51b7614f070a2a503cc977 Mon Sep 17 00:00:00 2001
From: Mathias Fiedler <mathias_fiedler@mentor.com>
Date: Mon, 17 Aug 2015 13:11:49 +0200
Subject: [PATCH 21/23] hack: reuse pushed buffers when running out of free
 buffers

This is a dirty workaround for releasing buffers which got dropped along
the pipeline.
---
 omx/gstomx.c | 58 ++++++++++++++++++++++++++++++++++++++++++++++------------
 omx/gstomx.h |  1 +
 2 files changed, 47 insertions(+), 12 deletions(-)

diff --git a/omx/gstomx.c b/omx/gstomx.c
index bcaad98..1fc6366 100644
--- a/omx/gstomx.c
+++ b/omx/gstomx.c
@@ -686,22 +686,25 @@ gboolean omx_port_out_buf_release_event(GstPad * pad,  GstObject * self, GstEven
       /* FIXME: What if the settings cookies don't match? */
       g_assert (ionbufp == ionbufpbuf->omx_buf->pAppPrivate);
 
-      ionbufp->used = TRUE;
+      if (g_queue_remove_all(&outport->pushed_buffers, ionbufp) > 0)
+      {
+        ionbufp->used = TRUE;
 
-      /* Reset all flags, some implementations don't
-       * reset them themselves and the flags are not
-       * valid anymore after the buffer was consumed
-       */
-      ionbufp->omx_buf->nFlags = 0;
+        /* Reset all flags, some implementations don't
+         * reset them themselves and the flags are not
+         * valid anymore after the buffer was consumed
+         */
+        ionbufp->omx_buf->nFlags = 0;
 
-      /* Reset offset and filled length */
-      ionbufp->omx_buf->nOffset = 0;
-      ionbufp->omx_buf->nFilledLen = 0;
+        /* Reset offset and filled length */
+        ionbufp->omx_buf->nOffset = 0;
+        ionbufp->omx_buf->nFilledLen = 0;
 
-      GST_DEBUG_OBJECT(self, "omx_port_out_buf_release_event: calling OMX_FillThisBuffer with GstOMXBuffer * %p and OMX_BUFFERHEADERTYPE %p", ionbufp, ionbufp->omx_buf);
+        GST_DEBUG_OBJECT(self, "omx_port_out_buf_release_event: calling OMX_FillThisBuffer with GstOMXBuffer * %p and OMX_BUFFERHEADERTYPE %p", ionbufp, ionbufp->omx_buf);
 
-      err = OMX_FillThisBuffer (comp->handle, ionbufp->omx_buf);
-      gst_omx_component_handle_messages (comp);
+        err = OMX_FillThisBuffer (comp->handle, ionbufp->omx_buf);
+        gst_omx_component_handle_messages (comp);
+      }
       g_mutex_unlock (&comp->lock);
 
       GST_DEBUG_OBJECT(self, "Recycled output buffer %p to %s port %u: error %s", ionbufp->omx_buf,  comp->name, outport->index, gst_omx_error_to_string (err));
@@ -1105,6 +1108,7 @@ gst_omx_component_add_port (GstOMXComponent * comp, guint32 index)
   port->port_def = port_def;
 
   g_queue_init (&port->pending_buffers);
+  g_queue_init (&port->pushed_buffers);
   port->flushing = TRUE;
   port->flushed = FALSE;
   port->enabled_pending = FALSE;
@@ -1492,6 +1496,31 @@ retry:
   if (g_queue_is_empty (&port->pending_buffers)) {
     GST_DEBUG_OBJECT (comp->parent, "Queue of %s port %u is empty",
         comp->name, port->index);
+
+    /* HACK: */
+    if ((_buf = g_queue_pop_head (&port->pushed_buffers)) != NULL)
+    {
+      _buf->used = TRUE;
+
+      /* Reset all flags, some implementations don't
+       * reset them themselves and the flags are not
+       * valid anymore after the buffer was consumed
+       */
+      _buf->omx_buf->nFlags = 0;
+
+      /* Reset offset and filled length */
+      _buf->omx_buf->nOffset = 0;
+      _buf->omx_buf->nFilledLen = 0;
+
+      GST_ERROR_OBJECT(comp->parent, "reusing unreleased GstOMXBuffer %p ", _buf);
+
+      OMX_FillThisBuffer (comp->handle, _buf->omx_buf);
+      // gst_omx_component_handle_messages (comp);
+
+      _buf = NULL;
+    }
+
+
     gst_omx_component_wait_message (comp, GST_CLOCK_TIME_NONE);
     gst_omx_component_handle_messages (comp);
 
@@ -1584,9 +1613,13 @@ gst_omx_port_release_buffer (GstOMXPort * port, GstOMXBuffer * buf)
   /* for Venus output buffer, wait for gstwaylandsink pad event to reuse buffer */
 #if defined (USE_OMX_TARGET_VENUS)
   if (port->port_def.eDir == OMX_DirOutput)
+  {
     GST_DEBUG_OBJECT(comp->parent,"Pushing to wayland: buf %p, buf->omx_buf %p, buf->omx_buf->pBuffer %p, nAllocLen %u, nFilledLen %u, nOffset %u",
 		     buf, buf->omx_buf, buf->omx_buf->pBuffer, (unsigned) buf->omx_buf->nAllocLen, (unsigned) buf->omx_buf->nFilledLen,
 		     (unsigned) buf->omx_buf->nOffset);
+
+    g_queue_push_tail (&port->pushed_buffers, buf);
+  }
   goto done;
 #else
   buf->used = TRUE;
@@ -1979,6 +2012,7 @@ gst_omx_port_deallocate_buffers_unlocked (GstOMXPort * port)
     g_slice_free (GstOMXBuffer, buf);
   }
   g_queue_clear (&port->pending_buffers);
+  g_queue_clear (&port->pushed_buffers);
   g_ptr_array_unref (port->buffers);
   port->buffers = NULL;
 
diff --git a/omx/gstomx.h b/omx/gstomx.h
index f13b696..525aff0 100644
--- a/omx/gstomx.h
+++ b/omx/gstomx.h
@@ -213,6 +213,7 @@ struct _GstOMXPort {
   OMX_PARAM_PORTDEFINITIONTYPE port_def;
   GPtrArray *buffers; /* Contains GstOMXBuffer* */
   GQueue pending_buffers; /* Contains GstOMXBuffer* */
+  GQueue pushed_buffers;
   gboolean flushing;
   gboolean flushed; /* TRUE after OMX_CommandFlush was done */
   gboolean enabled_pending;  /* TRUE after OMX_Command{En,Dis}able */
-- 
1.9.1

